### ISSUE TITLE
Issue with overlapping vector datatype

### ISSUE THREAD

**Issue by bosilca** [open-mpi/ompi#5540]
As reported by the DKRZ folks, creating a datatype with overlapping capabilities (legal for all send communications) leads to data corruption, as the packing and unpacking function will fail to compute the correct memory layout.

A simple example of such datatype is:

```c
MPI_Type_vector(3, 1, 0, MPI_INT, &btypes[0]);
MPI_Type_indexed(1, (int[]){1}, (int[]){0}, MPI_INT, &btypes[1]);

MPI_Type_create_struct(2, (int[]){1,1}, (MPI_Aint[]){0,12}, btypes, &dt);
```
All Open MPI versions are affected. However, it has been decided during the weekly call to only fix it in every release after 3.0 and obviously master.

**Comment by gpaulsen** [comment]
George has a patch that's coming.

**Comment by jsquyres** [comment]
Discussed on 2018-09-11 webex; @bosilca mentioned that he has a patch but he hasn't tested it properly with all the other datatype tests.  Stay tuned.

**Comment by gpaulsen** [comment]
Discussed at the RM meeting today, and will bring up again on tomorrow's Web-ex.

**Comment by jsquyres** [comment]
Per discussion on today's Webex, we decided that this issue should be fixed in the v4.0.x series, but probably isn't worth bringing back to the v3.0.x and v3.1.x series (i.e., it's a pretty esoteric issue for highly complex datatypes).

**Comment by gpaulsen** [comment]
@bosilca Do you have a PR?  If it's partial we might be able to get help working this...

**Comment by gpaulsen** [comment]
@bosilca ?

**Comment by jsquyres** [comment]
@bosilca said on the Webex today that he would separate out the fix for this vector issue from a branch where he has other optimizations, and then push a PR with just the fix.

**Comment by jsquyres** [comment]
@bosilca Any update on this?

**Comment by bosilca** [comment]
#6695 addresses this and much more.

**Comment by gpaulsen** [comment]
Added blocker label as this is affecting real app, and the fix PR #6695 can be picked to v4.0.2

**Comment by jsquyres** [comment]
@bosilca Notes that this does not easily cherry-pick to the v3.1.x or v3.0.x branches.  As such, @hppritcha and I agree that this will be a Known Issue on the v3.0.x / v3.1.x series: the solution will be to upgrade to >= v4.0.2.

**Comment by gpaulsen** [comment]
Re-opening this issue until it's been cherry-picked to v4.0.x

**Comment by gpaulsen** [comment]
v4.0.x backport: https://github.com/open-mpi/ompi/pull/6863

**Comment by gpaulsen** [comment]
The fix has been merged to both master and v4.0.x.  Closing this Issue.
